### ISSUE TITLE
Console Runner loads wrong .NET framework version when executing tests from multiple assemblies at once

### ISSUE THREAD

**Issue by PoByBolek** [nunit/nunit-console#667]
When running tests from multiple assemblies with different target frameworks, the console runner seems to only load the highest framework version.

For example, I have two test projects: one with target framework version 3.5 and one with target framework version 4.0.
The single test case in the 3.5 project looks like this:

    using System;
    using NUnit.Framework;

    namespace Test35
    {
        [TestFixture]
        public class TestFor35
        {
            [Test]
            public void Example()
            {
                Assert.AreEqual(new Version(2, 0, 0, 0), typeof(int).Assembly.GetName().Version);
            }
        }
    }

The single test case in the 4.0 project looks like this:

    using System;
    using NUnit.Framework;

    namespace Test40
    {
        [TestFixture]
        public class TestFor40
        {
            [Test]
            public void Example()
            {
                Assert.AreEqual(new Version(4, 0, 0, 0), typeof(int).Assembly.GetName().Version);
            }
        }
    }

Now, when I run the test assemblies individually, everything is fine:

    D:\temp\Nunit.Test>nunit3-console.exe --noresult Test35\bin\Debug\Test35.dll
    NUnit Console Runner 3.10.0 (.NET 2.0)
    Copyright (c) 2019 Charlie Poole, Rob Prouse
    Freitag, 9. August 2019 00:37:25

    Runtime Environment
       OS Version: Microsoft Windows NT 10.0.18362.0
      CLR Version: 4.0.30319.42000

    Test Files
        Test35\bin\Debug\Test35.dll

    Run Settings
        DisposeRunners: True
        WorkDirectory: D:\temp\Nunit.Test
        ImageRuntimeVersion: 2.0.50727
        ImageRequiresX86: False
        ImageRequiresDefaultAppDomainAssemblyResolver: False
        NumberOfTestWorkers: 4

    Test Run Summary
      Overall result: Passed
      Test Count: 1, Passed: 1, Failed: 0, Warnings: 0, Inconclusive: 0, Skipped: 0
      Start time: 2019-08-08 22:37:25Z
        End time: 2019-08-08 22:37:27Z
        Duration: 1.746 seconds

    D:\temp\Nunit.Test>nunit3-console.exe --noresult Test40\bin\Debug\Test40.dll
    NUnit Console Runner 3.10.0 (.NET 2.0)
    Copyright (c) 2019 Charlie Poole, Rob Prouse
    Freitag, 9. August 2019 00:37:41

    Runtime Environment
       OS Version: Microsoft Windows NT 10.0.18362.0
      CLR Version: 4.0.30319.42000

    Test Files
        Test40\bin\Debug\Test40.dll

    Run Settings
        DisposeRunners: True
        WorkDirectory: D:\temp\Nunit.Test
        ImageRuntimeVersion: 4.0.30319
        ImageTargetFrameworkName: .NETFramework,Version=v4.0
        ImageRequiresX86: False
        ImageRequiresDefaultAppDomainAssemblyResolver: False
        NumberOfTestWorkers: 4

    Test Run Summary
      Overall result: Passed
      Test Count: 1, Passed: 1, Failed: 0, Warnings: 0, Inconclusive: 0, Skipped: 0
      Start time: 2019-08-08 22:37:41Z
        End time: 2019-08-08 22:37:42Z
        Duration: 1.082 seconds

However, when I run both of them together, this happens:

    D:\temp\Nunit.Test>nunit3-console.exe --noresult Test35\bin\Debug\Test35.dll Test40\bin\Debug\Test40.dll
    NUnit Console Runner 3.10.0 (.NET 2.0)
    Copyright (c) 2019 Charlie Poole, Rob Prouse
    Freitag, 9. August 2019 00:42:16

    Runtime Environment
       OS Version: Microsoft Windows NT 10.0.18362.0
      CLR Version: 4.0.30319.42000

    Test Files
        Test35\bin\Debug\Test35.dll
        Test40\bin\Debug\Test40.dll

    Errors, Failures and Warnings

    1) Failed : Test35.TestFor35.Example
      Expected: <2.0.0.0>
      But was:  <4.0.0.0>
       bei Test35.TestFor35.Example() in D:\temp\Nunit.Test\Test35\TestFor35.cs:Zeile 12.

    Run Settings
        DisposeRunners: True
        WorkDirectory: D:\temp\Nunit.Test
        ImageRuntimeVersion: 2.0.50727
        ImageRequiresX86: False
        ImageRequiresDefaultAppDomainAssemblyResolver: False
        NumberOfTestWorkers: 4

    Test Run Summary
      Overall result: Failed
      Test Count: 2, Passed: 1, Failed: 1, Warnings: 0, Inconclusive: 0, Skipped: 0
        Failed Tests - Failures: 1, Errors: 0, Invalid: 0
      Start time: 2019-08-08 22:42:16Z
        End time: 2019-08-08 22:42:17Z
        Duration: 1.426 seconds

Interestingly, when I run them with `--domain=Multiple` (which should be the default according to the docs) this happens:

    D:\temp\Nunit.Test>nunit3-console.exe --domain=Multiple --noresult Test35\bin\Debug\Test35.dll Test40\bin\Debug\Test40.dll
    NUnit Console Runner 3.10.0 (.NET 2.0)
    Copyright (c) 2019 Charlie Poole, Rob Prouse
    Freitag, 9. August 2019 00:47:15

    Runtime Environment
       OS Version: Microsoft Windows NT 10.0.18362.0
      CLR Version: 4.0.30319.42000

    Test Files
        Test35\bin\Debug\Test35.dll
        Test40\bin\Debug\Test40.dll

    Test Run Summary
      Overall result: Unknown
      Test Count: 0, Passed: 0, Failed: 0, Warnings: 0, Inconclusive: 0, Skipped: 0
      Start time: 2019-08-08 22:47:16Z
        End time: 2019-08-08 22:47:16Z
        Duration: 0.477 seconds

**Comment by ChrisMaddock** [comment]
Interesting! Please can you add the logs for the failing case, by using —trace=Debug? Thanks!

**Comment by PoByBolek** [comment]
Sure, here you go.

[nunit-agent_19360.log](https://github.com/nunit/nunit-console/files/3485117/nunit-agent_19360.log)
[InternalTrace.4612.log](https://github.com/nunit/nunit-console/files/3485118/InternalTrace.4612.log)
[InternalTrace.16280.Test40.dll.log](https://github.com/nunit/nunit-console/files/3485119/InternalTrace.16280.Test40.dll.log)
[InternalTrace.19360.Test35.dll.log](https://github.com/nunit/nunit-console/files/3485120/InternalTrace.19360.Test35.dll.log)
[nunit-agent_16280.log](https://github.com/nunit/nunit-console/files/3485121/nunit-agent_16280.log)

**Comment by ChrisMaddock** [comment]
Thanks! I'm intrigued...

```
08:56:32.903 Debug [ 1] RuntimeFrameworkService: Assembly D:\temp\Nunit.Test\Test35\bin\Debug\Test35.dll uses version 2.0.50727
08:56:32.952 Debug [ 1] RuntimeFrameworkService: Assembly D:\temp\Nunit.Test\Test35\bin\Debug\Test35.dll targets 
08:56:32.953 Debug [ 1] RuntimeFrameworkService: Assembly D:\temp\Nunit.Test\Test40\bin\Debug\Test40.dll uses version 4.0.30319
08:56:32.963 Debug [ 1] RuntimeFrameworkService: Assembly D:\temp\Nunit.Test\Test40\bin\Debug\Test40.dll targets .NETFramework,Version=v4.0
08:56:32.964 Debug [ 1] RuntimeFrameworkService: Current framework is net-4.5
08:56:32.964 Debug [ 1] RuntimeFrameworkService: Requested framework is any
08:56:32.965 Debug [ 1] RuntimeFrameworkService: Test will use net-4.0 framework
[...]
08:56:32.981 Info  [ 4] ProcessRunner: Running Test35.dll
08:56:32.981 Info  [ 5] ProcessRunner: Running Test40.dll
08:56:32.991 Info  [ 4] TestAgency: Getting standard agent for use under net-4.5
08:56:32.991 Info  [ 5] TestAgency: Getting standard agent for use under net-4.5
```

I'll need to have a dig into the code to see what's going on here, but you've tickled my interest! Last request, mind pushing your Nunit.Test project to a GitHub repo, save me creating my own? Thanks! 😄

**Comment by CharliePoole** [comment]
Hmm... "Test35.dll targets"' nothing?

**Comment by PoByBolek** [comment]
Here's the sample repository: https://github.com/PoByBolek/Nunit.Test

**Comment by CharliePoole** [comment]
Your project file `Test35.csproj` has a `TargetFrameworkVersion` property on the Debug config only. This is also the case for `Test40.csproj` but that's not (yet) causing you a problem because the default of .NET 4.5 works OK for you.

It looks to me as if somebody editing the two `.csproj` files has mistakenly added a `Condition` to the first property group, which is clearly intended to contain properties available to both configs.

**Comment by PoByBolek** [comment]
No, the first `PropertyGroup` has no condition. It just sets the default configuration and platform (those tags have a condition attribute) when they are not provided. These are actually the default project files as Visual Studio creates them (except that I moved the duplicate `ErrorReport` and `WarningLevel` elements from the conditional `PropertyGroup`s to the unconditional (first) one.

**Comment by CharliePoole** [comment]
Yes, I see now you are right. I was hoping for a simple answer!

In any cases, it seems significant to me that the log for the 4.0 run displays the target framework, while that for the 3.5 run does not. The code in `RuntimeFrameworkService` is for some reason not finding the target framework when it inspects the assembly.

**Comment by PoByBolek** [comment]
Because the Test35 assembly does not have a [`TargetFrameworkAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.versioning.targetframeworkattribute?view=netframework-4.8) applied to it. And this is because this attribute only exists from .NET 4.0 onwards.

**Comment by ChrisMaddock** [comment]
Ok, had a quick look and confirmed the problem.

The RuntimeFrameworkService currently examines all sub-packages and determines the target framework for the top-level package, based as the max requirement for all subpackages.

It doesn't however define the required framework for any subpackages, so these are instead found to have no required framework, and run in default.

Have a simple fix, but maybe want to do a little refactoring around this. 🤔

**Comment by ChrisMaddock** [comment]
Fixed by #670
